CRM/Contact/BAO/RelationshipVortex.php

```diff
@@ -0,0 +1,112 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * Class CRM_Contact_BAO_RelationshipVortex.
+ */
+class CRM_Contact_BAO_RelationshipVortex extends CRM_Contact_DAO_RelationshipVortex {
+
+  /**
+   * Add our list of triggers to the global list.
+   *
+   * @param \Civi\Core\Event\GenericHookEvent $e
+   * @see \CRM_Utils_Hook::triggerInfo
+   */
+  public static function onHookTriggerInfo($e) {
+    $triggers = [
+      [
+        /**
+         * This trigger runs whenever a "civicrm_relationship" record is inserted or updated.
+         *
+         * Goal: Ensure that every relationship record has two corresponding entries in the
+         * vortex, the forward relationship (A=>B) and reverse relationship (B=>A).
+         */
+
+        'table' => 'civicrm_relationship',
+        'when' => 'AFTER',
+        'event' => ['INSERT', 'UPDATE'],
+        'sql' => "
+          DECLARE name_a_b_ VARCHAR(64);
+          DECLARE name_b_a_ VARCHAR(64);
+  
+          SELECT name_a_b, name_b_a INTO name_a_b_, name_b_a_ FROM civicrm_relationship_type WHERE id = NEW.relationship_type_id;
+  
+          INSERT INTO civicrm_relationship_vtx (relationship_id, relationship_type_id, orientation, near_contact_id, relation, far_contact_id)
+          VALUES (NEW.id, NEW.relationship_type_id, 0, NEW.contact_id_a, name_a_b_, NEW.contact_id_b)
+          ON DUPLICATE KEY UPDATE near_contact_id = NEW.contact_id_a, relation = name_a_b_, far_contact_id = NEW.contact_id_b;
+  
+          INSERT INTO civicrm_relationship_vtx (relationship_id, relationship_type_id, orientation, near_contact_id, relation, far_contact_id)
+          VALUES (NEW.id, NEW.relationship_type_id, 1, NEW.contact_id_b, name_b_a_, NEW.contact_id_a)
+          ON DUPLICATE KEY UPDATE near_contact_id = NEW.contact_id_b, relation = name_b_a_, far_contact_id = NEW.contact_id_a;
+        ",
+      ],
+
+      [
+        /**
+         * This trigger runs whenever a "civicrm_relationship_type" record is updated.
+         *
+         * Goal: Ensure that the denormalized fields ("name_b_a"/"name_a_b" <=> "type") remain current.
+         */
```

Comment:
Yea I guess it's technically possible to update a relationship type name, if you force your way past the UI (which definitely doesn't allow that) and use sql. Should be very rare, but ok.

Reply:
Yeah, it's rare, but it's more correct, and the marginal cost is very different - e.g. plugging this particular gap took maybe +15 min upfront; but any/each person who stumbles into the edge-case is liable to spend (optimistically) an hour figuring out that their UI is reporting weird data because somebody last week made an oversight during some special import/correction.